substrate/client/rpc-servers/src/middleware/chain_head.rs

```diff
@@ -0,0 +1,237 @@
+
+use std::{
+	collections::HashSet,
+	future::Future,
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+};
+
+use jsonrpsee::{
+	server::middleware::rpc::RpcServiceT,
+	types::{Params, Request},
+	MethodResponse,
+};
+use parking_lot::Mutex;
+use pin_project::pin_project;
+
+/// The per connectin data needed to manage chainHead subscriptions.
+#[derive(Default)]
+pub struct ConnectionData {
+	/// Active `chainHead_follow` subscriptions for this connection.
+	subscriptions: HashSet<String>,
+}
+
+/// Layer to allow the `chainHead` RPC methods to be called from a single connection.
+#[derive(Clone)]
+pub struct ChainHeadLayer {
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl ChainHeadLayer {
+	/// Create a new [`ChainHeadLayer`].
+	pub fn new(connection_data: Arc<Mutex<ConnectionData>>) -> Self {
+		Self { connection_data }
+	}
+}
+
+impl<S> tower::Layer<S> for ChainHeadLayer {
+	type Service = ChainHeadMiddleware<S>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		ChainHeadMiddleware::new(inner, self.connection_data.clone())
+	}
+}
+
+/// Chain head middleware.
+#[derive(Clone)]
+pub struct ChainHeadMiddleware<S> {
+	service: S,
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl<S> ChainHeadMiddleware<S> {
+	/// Create a new chain head middleware.
+	pub fn new(service: S, connection_data: Arc<Mutex<ConnectionData>>) -> ChainHeadMiddleware<S> {
+		ChainHeadMiddleware { service, connection_data }
+	}
+}
+
+impl<'a, S> RpcServiceT<'a> for ChainHeadMiddleware<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	type Future = ResponseFuture<S::Future>;
+
+	fn call(&self, req: Request<'a>) -> Self::Future {
+		const CHAIN_HEAD_FOLLOW: &str = "chainHead_unstable_follow";
+		const CHAIN_HEAD_CALL_METHODS: [&str; 8] = [
+			"chainHead_unstable_body",
+			"chainHead_unstable_header",
+			"chainHead_unstable_call",
+			"chainHead_unstable_unpin",
+			"chainHead_unstable_continue",
+			"chainHead_unstable_storage",
+			"chainHead_unstable_stopOperation",
+			"chainHead_unstable_unfollow",
+		];
+
+		let method_name = req.method_name();
+
+		// Intercept the subscription ID returned by the `chainHead_follow` method.
+		if method_name == CHAIN_HEAD_FOLLOW {
+			return ResponseFuture::Register {
+				fut: self.service.call(req.clone()),
+				connection_data: self.connection_data.clone(),
+			}
+		}
+
+		// Ensure the subscription ID of those methods corresponds to a subscription ID
+		// of this connection.
+		if CHAIN_HEAD_CALL_METHODS.contains(&method_name) {
+			let params = req.params();
+			let follow_subscription = get_subscription_id(params);
+
+			if let Some(follow_subscription) = follow_subscription {
+				if !self.connection_data.lock().subscriptions.contains(&follow_subscription) {
+					return ResponseFuture::Ready {
+						response: Some(MethodResponse::error(
+							req.id(),
+							jsonrpsee::types::error::ErrorObject::owned(
+								-32602,
+								"Invalid subscription ID",
```

Comment:
> A JSON-RPC error with error code -32602 is generated if one of the parameters doesn't correspond to the expected type (similarly to a missing parameter or an invalid parameter type).

I would probably add another error for this, the subscription ID is valid type but the subscription is not "active" on the connection....